File: src/main/java/com/arnas/klatrebackend/features/routes/RouteServiceDefault.java
package com.arnas.klatrebackend.features.routes;

import com.arnas.klatrebackend.features.auth.RequireGroupAccess;
import com.arnas.klatrebackend.features.auth.GroupAccessSource;
import com.arnas.klatrebackend.features.auth.Role;
import com.arnas.klatrebackend.features.images.ImageService;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Service;

@Service
public class RouteServiceDefault implements RouteService {

    private final RouteRepository routeRepository;
    private final ImageService imageService;

    public RouteServiceDefault(
            RouteRepository routeRepository,
            ImageService imageService
    ) {
        this.routeRepository = routeRepository;
        this.imageService = imageService;
    }

    @Override
    @RequireGroupAccess(minRole = Role.ADMIN, resolveGroupFrom = GroupAccessSource.FROM_ROUTE, sourceObjectParam = "routeDTO")
    public long addRoute(@NotNull RouteDTO routeDTO, long userId) {
        String imageId = null;
        if(routeDTO.getImage() != null)
            imageId = imageService.storeImageFile(routeDTO.getImage(), userId);
        return routeRepository.addRoute(routeDTO, imageId, userId);
    }

    @Override
    @RequireGroupAccess(minRole = Role.ADMIN, resolveGroupFrom = GroupAccessSource.FROM_ROUTE, sourceObjectParam = "routeDTO")
    public void updateRoute(@NotNull RouteUpdateDTO routeDTO, long userId) {
        var oldRouteOptional = routeRepository.getRouteById(routeDTO.getRouteId());
        if(oldRouteOptional.isEmpty())
            throw new RuntimeException("Route not found");
        var oldRoute = oldRouteOptional.get();
        String newImageId = null;
        if(routeDTO.getImage() != null) {
            if(oldRoute.getImageId() != null)
                imageService.deleteImage(oldRoute.getImageId());
            newImageId = imageService.storeImageFile(routeDTO.getImage(), userId);
        }

        var newRoute = new Route(
                routeDTO.getRouteId(),
                routeDTO.getName() != null ? routeDTO.getName() : oldRoute.getName(),
                routeDTO.getGradeId() != null ? routeDTO.getGradeId() : oldRoute.getGradeId(),
                routeDTO.getPlaceId() != null ? routeDTO.getPlaceId() : oldRoute.getPlaceId(),
                routeDTO.getDescription() != null ? routeDTO.getDescription() : oldRoute.getDescription(),
                routeDTO.getActive() != null ? routeDTO.getActive() : oldRoute.getActive(),
                routeDTO.getImage() != null ? newImageId : oldRoute.getImageId()
        );
        var rowAffected = routeRepository.updateRoute(newRoute);
        if(rowAffected <= 0) throw new RuntimeException("Failed to update boulder");

    }

    @Override
    @RequireGroupAccess(minRole = Role.ADMIN, resolveGroupFrom = GroupAccessSource.FROM_ROUTE)
    public void deleteRoute(long routeId) {
        var routeOptional= routeRepository.getRouteById(routeId);
        if(routeOptional.isEmpty()) throw new RuntimeException("Route not found");
        var route = routeOptional.get();
        if(route.getImageId() != null)
            imageService.deleteImage(route.getImageId());
        routeRepository.deleteRoute(routeId);
    }

    @Override
    @NotNull
    public RouteResponse getRoutesByPlace(long placeId, int page, int limit) {
        var pagingEnabled = limit > 0;
        var routes = routeRepository.getRoutesByPlace(placeId, page, limit + 1, pagingEnabled);
        return new RouteResponse(
                routes,
                page,
                limit,
                routeRepository.getNumRoutesInPlace(placeId, true),
                routeRepository.getNumRoutesInPlace(placeId, false),
                routes.size() > limit
        );
    }
}

File: src/main/kotlin/com/arnas/klatrebackend/features/routes/RouteServiceDefault.kt
package com.arnas.klatrebackend.features.routes

import com.arnas.klatrebackend.features.auth.RequireGroupAccess
import com.arnas.klatrebackend.features.auth.GroupAccessSource
import com.arnas.klatrebackend.features.auth.Role
import com.arnas.klatrebackend.features.images.ImageService
import org.springframework.stereotype.Service

@Service
class RouteServiceDefault(
    private val routeRepository: RouteRepository,
    private val imageService: ImageService
) : RouteService {

    @RequireGroupAccess(minRole = Role.ADMIN, resolveGroupFrom = GroupAccessSource.FROM_ROUTE, sourceObjectParam = "routeDTO")
    override fun addRoute(routeDTO: RouteDTO, userId: Long): Long {
        val image = routeDTO.image
        val imageId: String? = if (image != null) {
            imageService.storeImageFile(image, userId)
        } else null
        return routeRepository.addRoute(routeDTO, imageId, userId)
    }

    @RequireGroupAccess(minRole = Role.ADMIN, resolveGroupFrom = GroupAccessSource.FROM_ROUTE, sourceObjectParam = "routeDTO")
    override fun updateRoute(routeDTO: RouteUpdateDTO, userId: Long) {
        val oldRoute = routeRepository.getRouteById(routeDTO.routeId).orElseThrow {
            RuntimeException("Route not found")
        }
        val image = routeDTO.image
        var newImageId: String? = null
        if (image != null) {
            val oldImageId = oldRoute.imageId
            if (oldImageId != null) {
                imageService.deleteImage(oldImageId)
            }
            newImageId = imageService.storeImageFile(image, userId)
        }

        val newRoute = Route(
            id = routeDTO.routeId,
            name = routeDTO.name ?: oldRoute.name,
            gradeId = routeDTO.gradeId ?: oldRoute.gradeId,
            placeId = routeDTO.placeId ?: oldRoute.placeId,
            description = routeDTO.description ?: oldRoute.description,
            active = routeDTO.active ?: oldRoute.active,
            imageId = if (image != null) newImageId else oldRoute.imageId
        )
        val rowAffected = routeRepository.updateRoute(newRoute)
        if (rowAffected <= 0) throw RuntimeException("Failed to update boulder")
    }

    @RequireGroupAccess(minRole = Role.ADMIN, resolveGroupFrom = GroupAccessSource.FROM_ROUTE)
    override fun deleteRoute(routeId: Long) {
        val route = routeRepository.getRouteById(routeId).orElseThrow {
            RuntimeException("Route not found")
        }
        val imageId = route.imageId
        if (imageId != null) {
            imageService.deleteImage(imageId)
        }
        routeRepository.deleteRoute(routeId)
    }

    override fun getRoutesByPlace(placeId: Long, page: Int, limit: Int): RouteResponse {
        val pagingEnabled = limit > 0
        val routes = routeRepository.getRoutesByPlace(placeId, page, limit + 1, pagingEnabled)
        return RouteResponse(
            boulders = routes,
            page = page,
            limit = limit,
            activeBouldersCount = routeRepository.getNumRoutesInPlace(placeId, true),
            retiredBouldersCount = routeRepository.getNumRoutesInPlace(placeId, false),
            hasMore = routes.size > limit
        )
    }
}

